Proposals/20200511-HVPA-design.md

```diff
@@ -0,0 +1,129 @@
+maximum memory: 7G
+minimum replicas: 2
+maximum replicas: 5
+```
+For this bucket mimimum and maximum total CPU are 6 (= 3 * 2) and 30 (= 6 * 5) respectively, and same for the memory would be 5G and 35G respectively.
+So a configuration with 3 replicas with 6GB memory each would also lie in the same bucket.
+
+HVPA controller would calculate the total memory and CPU recommended by the recommenders (eg. HPA or VPA) by multiplying pod resources and the number of replicas, and would find the bucket for the calculated total resources. HVPA controller would then divide the bucket linearly to calculate the actual `replicas` and `resources` to which it will scale the application such that the total allocated resources to all its replicas is same as recommeded by the recommenders.
+
+### User Stories (optional)
+
+#### Story 1
+A user wants to scale both horizontally and vertically with folloing constraints:
+1. minimum replicas: 1, maximum replicas: 8
+1. minimum memory: 500M, maximum memory: 25G
+1. mimimum CPU: 0m, maximum CPU: 24000m
+
+The above requirement can be denoted as a single bucket.
+HVPA controller would divide the "CPU sub-bucket" and the "memory sub-bucket" linearly among the replica counts such that:
+1. each CPU bucket is of same size, ie, `3000m`
+1. each memory bucket is of same size, ie, `(10G - 100M) / 8`or `1.18G`
+
+Following is an image describing the divided CPU bucket. The numbers in the shaded boxes denote the maximum total resources that can be allocated for a replica count.
+![CPU Story 1](./images/CPUStory1.png)
+
+HVPA controller will take the total cpu and memory resources recommended by HPA or VPA, and will scale the application both horizontally and vertically such that:
+1. the total resoureces allocated to application is same as that recommended by HPA or VPA
+1. the total allocated resource is not more than the maximum total resource for that replica count. This way the resource per replica can be capped to a certain maximum value
+
+For example: If an application is running with 2 replicas with 6 cores each, and VPA and HPA recommendations are 12 cores and 4 replicas respectively, then HVPA would scale the application to 3 replicas and 8 cores. This is because both HPA and VPA recommend 24 cores, and this total lies in the 3rd column of 3 replicas. 
+
+#### Story 2
+A user wants to scale his/her application such that:
+1. It scales only vertically with 1 replica till it reaches 1 core and 2GB memory
+1. When more than 2GB memory is required, then the application should be scaled horizontally and vertically till memory reaches 20GB, and CPU reaches 9 cores, until maximum replica count of 8.
+
+In this case, user can provide 2 buckets:
+1. First bucket:
+   * Replicas: minimum: 1, maximum: 1
+   * CPU: minimum: 0, maximum: 1000m
+   * Memory: minimum: 0, maximum: 2G
+1. Second bucket:
+   * Replicas: minimum: 2, maximum: 8
+   * CPU: minimum: 1, maximum: 9000m
+   * Memory: minimum: 2G, maximum: 20G
+
```

Comment:
This is different from the API server discussion, but more importantly, seems to create problems, e.g. what happens around 1GB? What happens at all when going from 1.) bucket with a pod with 0.9 cpus to 1.1 cpus recommendation that brings you to 2 pods at 1.1 cpus or 0.55 cpus? But the latter would be below the min for 2 replicas, so in no-mans land. Even if the vertical scaling is not corrected, VPA would quickly make this recommendation and things, without hysteresis, would quickly flip back and forth. Scaling down is even more troubling as it needs correction, so 2 pods with 0.9 recommendation would bring you to 1 pod, but the cpu should be corrected for, i.e. 1.8 cpus, which doesn't fit into the first bucket.

We have solved all these problems in the last discussion and now it looks as if we start anew?

Reply:
Yes Vedran, I have corrected this in the PR. This is already taken care of in the code framework, however, skipped it while writing this doc. Sorry, my bad!

Reply:
So, we don't describe it anymore, but why then this document? The image also doesn't help me understand and seems to violate the statement.

On second pass, I believe to understand how you want to calculate it, but how that helps with hysteresis or how that fits the image, I don't see (yet).